### PR TITLE
Add select operation to kernel kenerator

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
Adds selection operation to kernel generator. It is equivalent to member function `.select()` on Eigen objects.

## Tests
Tests include one new reference kernel.

## Side Effects
None.

## Checklist

- [x] Math issue #1342 

- [x] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: test/unit/math/opencl/kernel_generator/select_test.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  EXPECT_EQ(expected_kernel_src, kernel_src);
+
+  res_cl = tmp;
+  MatrixXd res = stan::math::from_matrix_cl(res_cl);
+
+  MatrixXd correct = m_cond.select(m1, m2);
+  EXPECT_MATRIX_NEAR(res, correct, 1e-9);
+}
+
+TEST(MathMatrixCL, multiple_operations) {
```
Comment (rok-cesnovar) [inline test/unit/math/opencl/kernel_generator/select_test.cpp]: Mind also adding a zero-size test here? Something like: 
```cpp
TEST(MathMatrixCL, multiple_operations) {
  using stan::math::select;
  MatrixXd m0(0, 0);
  MatrixXd m1(2, 3);
  m1 << 1, 2.5, 3, 4, 5, 6.3;
  MatrixXi m2(2, 3);
  m2 << 10, 100, 1000, 30, 12, 0;
  MatrixXd m3(2, 3);
  m3 << 1.4, 12.5, 6.3, 34.4, -5, -6.3;
  Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> m_cond0(0, 0);
  Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> m_cond(2, 3);
  m_cond << true, false, false, true, true, false;
  Matrix<bool, Eigen::Dynamic, Eigen::Dynamic> m_cond2(2, 3);
  m_cond2 << false, false, true, true, false, true;

  matrix_cl<double> m0_cl(m0);
  matrix_cl<double> m1_cl(m1);
  matrix_cl<int> m2_cl(m2);
  matrix_cl<double> m3_cl(m3);
  matrix_cl<bool> m_cond_cl(m_cond);
  matrix_cl<bool> m_cond2_cl(m_cond2);
  matrix_cl<bool> m_cond0_cl(m_cond0);
  auto tmp = select(m_cond_cl, m1_cl, select(m_cond2_cl, m2_cl, m3_cl));
  matrix_cl<double> res_cl = tmp;

  MatrixXd res = stan::math::from_matrix_cl(res_cl);

  MatrixXd correct = m_cond.select(m1, m_cond2.select(m2.cast<double>(), m3));
  EXPECT_MATRIX_NEAR(res, correct, 1e-9);
  
  auto tmp0 = select(m_cond0_cl, m0_cl, select(m_cond0_cl, m0_cl, m0_cl));
  matrix_cl<double> res0_cl = tmp0;

  MatrixXd res0 = stan::math::from_matrix_cl(res0_cl);

  MatrixXd correct0 = m_cond0.select(m0, m_cond0.select(m0.cast<double>(), m0));
  EXPECT_MATRIX_NEAR(res0, correct0, 1e-9);
}
```

Comment (t4c1) [reply test/unit/math/opencl/kernel_generator/select_test.cpp]: I simplified your suggestion and put it into a separate test.

Comment (rok-cesnovar) [reply test/unit/math/opencl/kernel_generator/select_test.cpp]: Great.

**Comment by rok-cesnovar** [comment]
This looks good. I will approve and merge this after the release.

File: stan/math/opencl/kernel_generator/select.hpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+#include <set>
+#include <string>
+#include <type_traits>
+#include <utility>
+
+namespace stan {
+namespace math {
+
+/**
+ * Represents a selection operation in kernel generator expressions.
```
Comment (SteveBronder) [inline stan/math/opencl/kernel_generator/select.hpp]: [bikeshedding]

This is a little too broad, why not just say "Represents an ternary matrix operation similar to Eigen's select statement"

File: stan/math/opencl/kernel_generator/select.hpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                       "rows of ", "else", els.rows());
+    }
+    if (condition.cols() != base::dynamic && els.cols() != base::dynamic) {
+      check_size_match("select", "Columns of ", "condition", condition.cols(),
+                       "columns of ", "else", els.cols());
+    }
+  }
+
+  /**
+   * generates kernel code for this expression.
```
Comment (SteveBronder) [inline stan/math/opencl/kernel_generator/select.hpp]:
```suggestion
   * generates kernel code for a ternary operation
```

File: stan/math/opencl/kernel_generator/select.hpp (hunk 2 above)
Comment (SteveBronder) [reply stan/math/opencl/kernel_generator/select.hpp]: Or like what you have below

>  Selection operation on kernel generator expressions. This is element wise ternary operator <code> condition ? then : els </code>.

File: stan/math/opencl/kernel_generator/select.hpp (hunk 3 above)
Comment (t4c1) [reply stan/math/opencl/kernel_generator/select.hpp]: I would prefere to leave this one as it is.

File: stan/math/opencl/kernel_generator/select.hpp (hunk 2 above)
Comment (t4c1) [reply stan/math/opencl/kernel_generator/select.hpp]: done

File: stan/math/opencl/kernel_generator/select.hpp (hunk 3 above)
Comment (rok-cesnovar) [reply stan/math/opencl/kernel_generator/select.hpp]: What about just going with "for the select operation"?

Comment (t4c1) [reply stan/math/opencl/kernel_generator/select.hpp]: done

**Pull request merged**